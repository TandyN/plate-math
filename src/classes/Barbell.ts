import Weight from './Weight'

class Barbell {
  barbell_weight: number
  weights: Array<Weight>

  constructor(barbell_weight = 20) {
    this.barbell_weight = barbell_weight
    this.weights = []
  }
}
