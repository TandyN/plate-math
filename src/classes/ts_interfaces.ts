enum WeightType {
  kgs = 'kilograms',
  lbs = 'pounds',
}

export { WeightType }
