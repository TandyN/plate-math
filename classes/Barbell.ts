import { Plate } from './Plate'
import { WeightType } from './ts_interfaces'

class Barbell {
  weight: number // will always be in kilograms
  attached_weights: Array<Plate>

  constructor(weight = 20, weight_type = WeightType.kgs) {
    if (weight_type === WeightType.lbs) {
      this.weight = weight / 2.205
    } else {
      this.weight = weight
    }
    this.attached_weights = []
  }

  add_kilogram_plate(weight: number): Plate {
    const added_kilogram_plate: Plate = new Plate(weight, WeightType.kgs)
    this.attached_weights.push(added_kilogram_plate)
    return added_kilogram_plate
  }

  add_pound_plate(weight: number): Plate {
    const added_pound_plate: Plate = new Plate(weight, WeightType.lbs)
    this.attached_weights.push(added_pound_plate)
    return added_pound_plate
  }

  remove_last_plate() {
    const removed_plate: Plate | undefined = this.attached_weights.pop()
    return removed_plate
  }

  remove_all_plates() {
    const current_attached_plates: Array<Plate> = this.attached_weights
    this.attached_weights = []
    return current_attached_plates
  }

  get_total_kilograms(): number {
    return (
      this.attached_weights.reduce((current_total, weight_node): number => {
        let current_weight: number = weight_node.weight
        if (weight_node.weight_type === WeightType.lbs) {
          current_weight /= 2.205
        }
        return current_total + current_weight
      }, 0) *
        2 +
      this.weight
    )
  }

  get_total_pounds(): number {
    return (
      this.attached_weights.reduce((current_total, weight_node): number => {
        let current_weight: number = weight_node.weight
        if (weight_node.weight_type === WeightType.kgs) {
          current_weight *= 2.205
        }
        return current_total + current_weight
      }, 0) *
        2 +
      this.weight * 2.205
    )
  }

  fill_with_kilogram_plates(
    weight: number,
    weight_type: WeightType = WeightType.kgs,
  ): Array<Plate> {
    let target_weight: number

    if (weight_type === WeightType.lbs) {
      target_weight = weight / 2.205
    } else {
      target_weight = weight
    }

    const target_attached_weights: Array<Plate> = []

    const kilogram_plates: Array<number> = [
      25, 20, 15, 10, 5, 2.5, 1.25, 0.5, 0.25,
    ]

    let current_weight: number = this.get_total_kilograms()
    let plate_index = 0

    while (plate_index < kilogram_plates.length) {
      const added_weight = current_weight + kilogram_plates[plate_index] * 2

      if (added_weight <= target_weight) {
        current_weight = added_weight
        target_attached_weights.push(
          new Plate(kilogram_plates[plate_index], WeightType.kgs),
        )
        this.attached_weights.push(
          new Plate(kilogram_plates[plate_index], WeightType.kgs),
        )
      } else {
        plate_index++
      }
    }

    return target_attached_weights
  }

  fill_with_pound_plates(
    weight: number,
    weight_type: WeightType = WeightType.kgs,
  ): Array<Plate> {
    let target_weight: number

    if (weight_type === WeightType.kgs) {
      target_weight = weight * 2.205
    } else {
      target_weight = weight
    }

    const target_attached_weights: Array<Plate> = []

    const pound_plates: Array<number> = [45, 35, 25, 10, 5, 2.5, 1.25]

    let current_weight: number = this.get_total_pounds()
    let plate_index = 0

    while (plate_index < pound_plates.length) {
      const added_weight = current_weight + pound_plates[plate_index] * 2

      if (added_weight <= target_weight) {
        current_weight = added_weight
        target_attached_weights.push(
          new Plate(pound_plates[plate_index], WeightType.lbs),
        )
        this.attached_weights.push(
          new Plate(pound_plates[plate_index], WeightType.lbs),
        )
      } else {
        plate_index++
      }
    }

    return target_attached_weights
  }
}

export { Barbell }
