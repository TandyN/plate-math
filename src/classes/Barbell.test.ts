import { Barbell } from './Barbell'
import { WeightType } from './ts_interfaces'

describe('Barbell Class', () => {
  it('should convert all weight to kilograms if different weight type is specified', () => {
    const barbell = new Barbell(45, WeightType.lbs)
    expect(barbell.weight).toBe(45 / 2.205)
  })

  it('should assume barbell weight type is kilograms if weight type does not exist', () => {
    const barbell = new Barbell(33, 'fake weight type' as never)
    expect(barbell.weight).toBe(33)
  })

  it('should default barbell weight to 20', () => {
    const barbell = new Barbell()
    expect(barbell.weight).toBe(20)
  })

  describe('add_kilogram_plate function', () => {
    it('should add to attached_weights if weight number exists as a kilogram plate', () => {
      const barbell = new Barbell()
      barbell.add_kilogram_plate(10)

      expect(barbell.attached_weights.length).toBe(1)
      expect(barbell.attached_weights[0].weight).toBe(10)
      expect(barbell.attached_weights[0].weight_type).toBe('kilograms')

      barbell.add_kilogram_plate(15)
      expect(barbell.attached_weights.length).toBe(2)
      expect(barbell.attached_weights[1].weight).toBe(15)
      expect(barbell.attached_weights[1].weight_type).toBe('kilograms')
    })

    it('should throw an error if number does not exist', () => {
      const barbell = new Barbell()
      expect(() => {
        barbell.add_kilogram_plate(11)
      }).toThrowErrorMatchingSnapshot()
    })
  })

  describe('add_pound_plate function', () => {
    it('should add to attached_weights if weight number exists as a pound plate', () => {
      const barbell = new Barbell()
      barbell.add_pound_plate(10)

      expect(barbell.attached_weights.length).toBe(1)
      expect(barbell.attached_weights[0].weight).toBe(10)
      expect(barbell.attached_weights[0].weight_type).toBe('pounds')

      barbell.add_pound_plate(25)
      expect(barbell.attached_weights.length).toBe(2)
      expect(barbell.attached_weights[1].weight).toBe(25)
      expect(barbell.attached_weights[1].weight_type).toBe('pounds')
    })

    it('should throw an error if number does not exist', () => {
      const barbell = new Barbell()
      expect(() => {
        barbell.add_pound_plate(11)
      }).toThrowErrorMatchingSnapshot()
    })
  })

  describe('remove_plate function', () => {
    it('should remove plate from attached_weights', () => {
      const barbell = new Barbell()
      barbell.add_pound_plate(10)
      barbell.add_pound_plate(10)
      expect(barbell.attached_weights.length).toBe(2)
      barbell.remove_plate()
      expect(barbell.attached_weights.length).toBe(1)
      barbell.remove_plate()
      expect(barbell.attached_weights.length).toBe(0)
    })
  })

  describe('get_total_kilograms function', () => {
    it('should calculate all plates attached to barbell in kilograms', () => {
      const barbell = new Barbell()
      barbell.add_pound_plate(10)
      barbell.add_kilogram_plate(10)
      barbell.add_pound_plate(10)
      barbell.add_kilogram_plate(10)
      barbell.add_pound_plate(5)

      const actual = Math.round(barbell.get_total_kilograms() * 100) / 100

      // add lb plates, multiple by 2 for both sides of the barbell, convert to kgs, add barbell weight, then add kg plates at the end for both sides
      const expected = (25 * 2) / 2.205 + 20 + 40

      expect(actual).toBe(Math.round(expected * 100) / 100)
    })
  })

  describe('get_total_pounds function', () => {
    it('should calculate all plates attached to barbell in pounds', () => {
      const barbell = new Barbell()
      barbell.add_kilogram_plate(10)
      barbell.add_pound_plate(10)
      barbell.add_kilogram_plate(10)
      barbell.add_pound_plate(10)
      barbell.add_kilogram_plate(5)

      const actual = Math.round(barbell.get_total_pounds() * 100) / 100

      // add kg plates, multiple by 2 for both sides of the barbell, convert to lbs, add barbell weight, then add pound lb at the end for both sides
      const expected = 25 * 2 * 2.205 + 20 * 2.205 + 40

      expect(actual).toBe(Math.round(expected * 100) / 100)
    })
  })
})
