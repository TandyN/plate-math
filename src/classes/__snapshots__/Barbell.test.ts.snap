// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Barbell Class add_kilogram_plate function should throw an error if number does not exist 1`] = `"plate number '11' does not exist for 'kilograms'"`;

exports[`Barbell Class add_pound_plate function should throw an error if number does not exist 1`] = `"plate number '11' does not exist for 'pounds'"`;
